### ISSUE TITLE
API Proposal: Add a ValueStringBuilder

### ISSUE THREAD

**Issue by JeremyKuhne** [dotnet/corefx#28379]
We should consider making a value based StringBuilder to allow low allocation building of strings.  While Span allows you to provide a writable buffer, in many scenarios we have a need to get or build strings and we don't know precisely how much space will be needed ahead of time. Having an abstraction that can grow beyond a given initial buffer is particularly useful as it doesn't require looping with `Try*` APIs- which can be both complicated and have negative performance implications.

We currently use `ValueStringBuilder` for this purpose internally. It starts with an optional initial buffer (which we often stackalloc) and will grow using ArrayPool<char> if needed.

This is the current shape of our internal [`ValueStringBuilder`](https://github.com/dotnet/coreclr/blob/bb0e771d2d08fedce8fc598e620c5d66b361e9e7/src/mscorlib/shared/System/Text/ValueStringBuilder.cs):

``` C#
namespace System.Text
{
    internal ref struct ValueStringBuilder
    {
        public ValueStringBuilder(Span<char> initialBuffer);
        public int Length { get; set; }
        public int Capacity { get; }
        public void EnsureCapacity(int capacity);

        /// <summary>
        /// Get a pinnable reference to the builder.
        /// </summary>
        /// <param name="terminate">Ensures that the builder has a null char after <see cref="Length"/></param>
        public ref char GetPinnableReference(bool terminate = false);
        public ref char this[int index] { get; }

        // Returns a string based off of the current position
        public override string ToString();

        /// <summary>
        /// Returns a span around the contents of the builder.
        /// </summary>
        /// <param name="terminate">Ensures that the builder has a null char after <see cref="Length"/></param>
        public ReadOnlySpan<char> AsSpan(bool terminate);

        // To ensure inlining perf, we have a separate overload for terminate
        public ReadOnlySpan<char> AsSpan();

        public bool TryCopyTo(Span<char> destination, out int charsWritten);
        public void Insert(int index, char value, int count);
        public void Append(char c);
        public void Append(string s);
        public void Append(char c, int count);
        public unsafe void Append(char* value, int length);
        public void Append(ReadOnlySpan<char> value);

        // This gives you an appended span that you can write to
        public Span<char> AppendSpan(int length);

        // Returns any ArrayPool buffer that may have been rented
        public void Dispose()
    }
}
```

Here is a common pattern on an API that could theoretically be made public if ValueStringBuilder was public:
(Although we would call this one FullUserName or something like that.)

https://github.com/dotnet/corefx/blob/050bc33738887d9d8fcc9bc5965b7d9ca65bc7f4/src/System.Runtime.Extensions/src/System/Environment.Win32.cs#L40-L56

The caller is above this method:

https://github.com/dotnet/corefx/blob/050bc33738887d9d8fcc9bc5965b7d9ca65bc7f4/src/System.Runtime.Extensions/src/System/Environment.Win32.cs#L13-L38

I'll add more usage details and possible API surface area.

### Notes

- Does using this remove the need for having a `bool TryGet*(Span)` overload? (see https://github.com/dotnet/coreclr/pull/17097/files#r176560435)
- Should we allow you to make it non-growable? (So we can avoid TryGet above?)
- Can we get C# to allow using ref structs that have a `Dispose()` in a `using` statement?
- Do we care about `AppendFormat` overloads?

**Comment by JeremyKuhne** [comment]
@jkotas, @stephentoub, @KrzysztofCwalina, @danmosemsft, @ahsonkhan

**Comment by stephentoub** [comment]
> bool terminate

I didn't see that we'd added these arguments. I don't think we should have them.  If you want to null terminate, you can call `Append('\0')`.

**Comment by JeremyKuhne** [comment]
> I didn't see that we'd added these arguments. I don't think we should have them. If you want to null terminate, you can call Append('\0').

Discussed this one at length with @jkotas in https://github.com/dotnet/coreclr/pull/16921. The problem with appending a null is that it changes the `Length` and breaks the convention on what `Length` means for `strings` & `StringBuilder` (i.e. doesn't include the terminator).  It is incredibly error prone as a result (as my need for the mentioned pull demonstrates).

`StringBuilder` always terminates, but doing this comes at a cost. It is only really needed in interop calls that take a `sz` (most Win32).

**Comment by stephentoub** [comment]
> The problem with appending a null is that it changes the Length and breaks the convention on what Length means for strings & StringBuilder (i.e. doesn't include the terminator).

The `bool terminate` is also changing the Length.

**Comment by JeremyKuhne** [comment]
> The `bool terminate` is also changing the Length.

No, it leaves Length as is. It _might_ increase Capacity.

**Comment by stephentoub** [comment]
Hmm, I see, ok.

**Comment by omariom** [comment]
Are they needed if there is an overload taking ROSpan?

public void Append(string s);
public unsafe void Append(char* value, int length);

**Comment by JeremyKuhne** [comment]
> Are they needed if there is an overload taking ROSpan?

No, they aren't.  I'll add a proposed section.

**Comment by omariom** [comment]
> Can we get C# to allow using ref structs that have a Dispose() in a using statement?

There is a proposal https://github.com/dotnet/csharplang/issues/93
It would be even better with this one https://github.com/dotnet/csharplang/issues/114

**Comment by KrzysztofCwalina** [comment]
There are several related efforts in corfxlab. We should try to unify (or make them consistent). See:
https://github.com/dotnet/corefxlab/blob/master/src/System.Buffers.ReaderWriter/System/Buffers/Writer/BufferWriter.cs
https://github.com/dotnet/corefxlab/blob/master/src/System.Buffers.ReaderWriter/System/Buffers/Writer/BufferWriterT_ints.cs
https://github.com/dotnet/corefxlab/blob/master/src/System.Buffers.ReaderWriter/System/Buffers/Writer/BufferWriter_strings.cs
And here support for non-allocating composite formats: https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Formatting/System/Text/Formatting/CompositeFormat.cs

**Comment by JeremyKuhne** [comment]
@KrzysztofCwalina thanks for the links! I've updated the text with design goals and will review the other efforts.

**Comment by MarcoRossignoli** [comment]
>Should we allow you to make it non-growable? (So we can avoid TryGet above?)

>// some reason it isn't, we'll grow the buffer. 

In case of growable could be useful to have a boolean properties to know if VSB has grown? In a hot path could be useful avoid to add pressure to underlying pool(today) and improve performance by choosing the right size of initial buffer.

**Comment by Joe4evr** [comment]
Some questions, I guess...
> ```cs
> public void Insert(int index, char value, int count = 1);
> public void Insert(int index, ReadOnlySpan<char> value, int count = 1);
> public void Append(char c, int count = 1);
> public void Append(ReadOnlySpan<char> value);
> ```

* Are these the only methods to build with, or will there be more parity with the existing `StringBuilder` API (`AppendLine`, `string` parameter, et al.) and those are omitted from the post for brevity?
* Is there a reason these don't return the `this` instance to enable fluent building?

**Comment by khellang** [comment]
Speaking of related efforts... There's also the [`InplaceStringBuilder`](https://github.com/aspnet/Common/blob/dev/src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs) from Microsoft.Extensions.Primitives, which is used in MVC and the HTTP abstractions. Might want to sync up with them as well.

// @pakrym @Tratcher

**Comment by JeremyKuhne** [comment]
> In case of growable could be useful to have a boolean properties to know if VSB has grown? 

Perhaps we should add an EventSource log for this?

**Comment by JeremyKuhne** [comment]
@Joe4evr 

> Are these the only methods to build with, or will there be more parity with the existing StringBuilder API (AppendLine, string parameter, et al.) and those are omitted from the post for brevity?

These were the most critical/obvious (based on current usage). I didn't want to start the discussion with too much complexity. Attacking this in priority order seemed like it would go a little quicker.

> Is there a reason these don't return the this instance to enable fluent building?

Hadn't considered it is all. I'll take a look.

**Comment by jkotas** [comment]
> fluent building

Fluent pattern adds overhead, and it does not work for structs that needs to be disposed.

**Comment by Joe4evr** [comment]
> Fluent pattern adds overhead, and it does not work for structs that needs to be disposed.

Does it, now? Surely, it's less overhead than display class allocation or something.
![](https://i.imgur.com/Uxv8e1Y.png)

**Comment by kasper3** [comment]
Combining two of @khellang's comment https://github.com/dotnet/corefx/issues/20378#issuecomment-304400894 & https://github.com/dotnet/corefx/issues/28379#issuecomment-375975366, it would be super helpful for almost all the .NET developers if we get a comprehensive blog article or Channel 9 video on all these newly optimized string related types in .NET Core and ASP.NET Core detailing when to use which and why. During that exercise if some overlapping types are found, please consider deprecating one. With every upcoming string-related type, it is getting scarier. 😱

**Comment by khellang** [comment]
> it would be super helpful [...] if we get a comprehensive blog article or Channel 9 video on all these newly optimized string related types [...] detailing when to use which and why.

I think the general guidance (once `Span<T>` and friends officially hits the streets) is to use `Span<T>` (or `ReadOnlySpan<T>`, `Memory<T>`, `ReadOnlyMemory<T>` etc.) whenever you can. `StringSegment`, `ArraySegment<T>` and various other related types (some of which has been around for some time) are all efforts to achieve some of the same benefits as the System.Memory types.

**Comment by jnm2** [comment]
Obsolete? https://github.com/dotnet/corefx/issues/20378

**Comment by khellang** [comment]
No, but with the other new APIs, it's arguable how much value yet another part-or-string type would bring vs the overhead of the type existing at all. So I would say "redundant" 😉

**Comment by Joe4evr** [comment]
:confused: The API proposal itself clearly states:
> but spans have different type system constraints that not all consumers can -- or want to -- opt into.

**Comment by khellang** [comment]
I'm assuming @terrajobst is referring to the stack-only-ness of `ReadOnlySpan<T>`. That doesn't apply to `ReadOnlyMemory<T>`.

There might be valid reasons for yet another slice-of-string type, but as I mentioned in #20378, the number of these types are getting a bit out of hand and it's hard to know which to use and which to support in libraries.

**Comment by JeremyKuhne** [comment]
The key thing we're missing that this addresses is a low allocation expandable `Span<T>` provider. We could go super simple with something like `ref SpanProvider<T>` cutting all string-like methods and perhaps just providing the functionality as extensions. I'm not fundamentally opposed to going that route, but the fact is that we do provide `StringBuilder`. It is a concept that people are familiar with and we need a mechanism to replace it for performance sensitive code paths. It seems less burdensome to me to juggle `StringBuilder`/`ValueStringBuilder` than `StringBuilder`/`SpanProvider<char>`, particularly as we're already moving down the road of having `Value*` value type equivalents of other reference types (e.g. it is an emergent pattern).

**Comment by fanoI** [comment]
I have a question regarding the Dispose() of this structure it should be always called?
If so the posted examples are not risking to leaking memory?

**Comment by JeremyKuhne** [comment]
> I have a question regarding the Dispose() of this structure it should be always called?

It returns any potentially rented buffers to the array pool. `ToString()` internally also returns the buffer, I'll call that out. On a related note, we don't usually use these with `try ... finally` blocks as in many cases we don't expect throws to be common enough to warrant taking the performance hit of introducing the block- we're ok with not returning the buffer to the pool in those cases (which will get it garbage collected).

**Comment by JeremyKuhne** [comment]
Note that I'm exploring alternative options utilizing some of the concepts @KrzysztofCwalina brought up (with @KrzysztofCwalina). We'll circle back around with more info.

**Comment by ViktorHofer** [comment]
@JeremyKuhne have you made progress on exploring new options? I think we should definitely get this into 3.0.

**Comment by vancem** [comment]
+ @vancem

**Comment by iSazonov** [comment]
I'd expect that the API has a parity with StringBuilder:
- AppendFormat
- AppendJoin
- AppendLine
- Clear
- Remove
- Replace

**Comment by stephentoub** [comment]
> I'd expect that the API has a parity with StringBuilder

I don't believe that should be a goal.  They're meant for different uses and therefore don't need perfect symmetry.  For example, ValueStringBuilder is meant for situations where you're trying to minimize allocation overhead as much as possible, and in such situations, you're not going to want to use a method that takes a params object[].

**Comment by danmosemsft** [comment]
Not params methods, but Clear() is reasonable, and also Remove and Replace given it already has O(n) methods like Insert.

**Comment by danmosemsft** [comment]
AppendFormat() does a lot of work (and equally requires a lot of code to implement). This type is more narrowly targeted at scenarios where you want to accept a bit more complexity for the most possible performance, in that scenario you are likely to handle any formatting yourself.

**Comment by JeremyKuhne** [comment]
@iSazonov Having format methods on this in the future is definitely something I'm looking at. I'll be putting a prototype of one in corefxlab _very_ soon. Doing formatting in a non-allocating way is a little tricky given boxing, etc. As such it will take a little while to shake out and I don't want to hold up the core functionality here. I'll make sure to tag you on the prototype PR.

**Comment by jkotas** [comment]
`StringBuilder` is optimized for convenience and broad range of scenarios. It will continue to work better than `ValueStringBuilder` in many case.

`ValueStringBuilder` is not optimized for convenience. It is optimized for building small strings (<1kB) where the work required to format the string is small and fixed `StringBuilder` overhead is too high.

**Comment by iSazonov** [comment]
> where the work required to format the string is small 

Could you please clarify that do you mean? I'd expect that formatting is intended primarily for people and result string will definitely be smaller 1 Kb like in example I pointed above.

**Comment by jkotas** [comment]
Formatting strings and string interpolation as implemented in .NET today are convenience that you pay quite a bit for.

String formatting/interpolation that is both convenient to use *and* efficient as handwritten formatting would require C# compiler changes, I think.

**Comment by terrajobst** [comment]
> One possibility for forced by-ref structs that has been raised is `ref local`.

I'm sad it's not `ref ref` 😉

**Comment by Joe4evr** [comment]
Since the language team has expressed interest in allowing more features that might need new runtime constructs, it seems best to consider "Non-Copyable structs" on a list of features that may warrant future-CLR enforcement.
